lib/dataContract/stateTransition/validation/validateDataContractMaxDepthFactory.js

```diff
@@ -0,0 +1,82 @@
+const ValidationResult = require('../../../validation/ValidationResult');
+const JsonSchemaError = require('../../../errors/JsonSchemaError');
+const DataContractMaxDepthExceedError = require('../../../errors/DataContractMaxDepthExceedError');
+
+/**
+ * Check that JSON Schema max depth is less than max value
+ * @private
+ * @param {Object} json
+ * @returns {number}
+ */
+function checkMaxDepth(json) {
+  let depth = 1;
+
+  const keys = Object.keys(json);
+
+  for (let i = 0; i < keys.length; i++) {
+    const key = keys[i];
+
+    if (typeof json[key] === 'object') {
+      const tmpDepth = checkMaxDepth(json[key]) + 1;
+      depth = Math.max(depth, tmpDepth);
+    }
+  }
+
+  if (depth > DataContractMaxDepthExceedError.MAX_DEPTH) {
+    throw new DataContractMaxDepthExceedError();
+  }
+
+  return depth;
+}
+
+/**
+ *
+ * @param {$RefParser} refParser
+ * @returns {validateDataContractMaxDepth}
+ */
+function validateDataContractMaxDepthFactory(refParser) {
+  /**
+   * Dereference JSON Schema $ref pointers and check max depth
+   * @typedef validateDataContractMaxDepth
+   * @param {RawDataContract} dataContract
+   * @returns {ValidationResult}
+   */
+  async function validateDataContractMaxDepth(dataContract) {
+    const result = new ValidationResult();
+
+    let dereferencedDataContract;
+
+    try {
+      dereferencedDataContract = await refParser.dereference(dataContract, {
+        parse: {
+          yaml: false,
+          text: false,
+          binary: false,
+        },
+        resolve: {
+          external: false,
+          http: false,
+        },
+        dereference: {
+          circular: false,
+        },
+      });
+    } catch (error) {
+      result.addError(new JsonSchemaError(error));
+    }
+
+    if (dereferencedDataContract) {
+      try {
+        checkMaxDepth(dereferencedDataContract);
+      } catch (error) {
+        result.addError(error);
+      }
```

Comment:
If guess it doesn't make sense to check depth in case of ref error. Let's just return result here.